### PR TITLE
CLN: More numpy 2 stuff

### PR DESCRIPTION
**Pull request by lithomas1**
- [ ] closes #xxxx (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

**Review by WillAyd — APPROVED** [review]
nice work

File: pandas/_libs/src/vendored/ujson/python/objToJSON.c
Comment (WillAyd) [inline pandas/_libs/src/vendored/ujson/python/objToJSON.c:493]: The failing build doesn't like the implicit conversion from a PyArrayObject* to a PyObject* . Should be able to type-check / cast to make the compiler happy

File: pandas/_libs/src/vendored/ujson/python/objToJSON.c
Comment (WillAyd) [inline pandas/_libs/src/vendored/ujson/python/objToJSON.c:494]: Would still prefer a type check before the cast - prevents strange bugs in case this ever gets refactored.

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/_libs/src/vendored/ujson/python/objToJSON.c
Comment (lithomas1) [reply pandas/_libs/src/vendored/ujson/python/objToJSON.c:494]: Looks like it already happens a couple lines above.
https://github.com/pandas-dev/pandas/blob/e14a9bd41d8cd8ac52c5c958b735623fe0eae064/pandas/_libs/src/vendored/ujson/python/objToJSON.c#L477-L480

Comment (WillAyd) [reply pandas/_libs/src/vendored/ujson/python/objToJSON.c:494]: In that case we should assign `const PyArrayObject *arrayobj = (const PyArrayObject *)npyarr->array` right after the type check and reference that here.

Generally want to minimize casting and keep it very localized to where the type check happens (historically we have not done a good job of this) Otherwise things are very liable to break when refactoring

**Review by WillAyd — APPROVED** [review]
lgtm. didn't look through all failures but assume they are leap day issues. maybe worth waiting until tomorrow

**Comment by WillAyd** [comment]
I merged main thinking the CI failures were unrelated but looks like they are still around. Happy to merge as is given I am fairly confident the failures are unrelated, but will leave it up to you

**Comment by lithomas1** [comment]
Merging then, since numpy is blocked on this.

Thanks for the review!

**Pull request merged**
